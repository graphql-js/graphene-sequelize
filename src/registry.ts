import { assertSequelizeModel, Model } from "./utils";
import * as Sequelize from "sequelize";
import { getSequelizeModel } from "./reflection";

export type RegistryMapping = Sequelize.Model<any, any>[];

export class Registry {
  public mapping: RegistryMapping[];
  constructor() {
    this.mapping = [];
  }
  public register(sequelizeType: any) {
    let model: Model = getSequelizeModel(sequelizeType);
    this.mapping.push(sequelizeType);
  }
  public getTypeForModel(model: Sequelize.Model<any, any>) {
    assertSequelizeModel(model);
    for (let sequelizeType of this.mapping) {
      let sequelizeTypeModel: Model = getSequelizeModel(sequelizeType);
      if (sequelizeTypeModel === model) {
        return sequelizeType;
      }
    }
    return null;
  }
}

let registry: Registry | null = null;

export const getGlobalRegistry = (): Registry => {
  if (registry === null) {
    registry = new Registry();
  }
  return registry;
};

export const resetGlobalRegistry = () => {
  registry = null;
};
