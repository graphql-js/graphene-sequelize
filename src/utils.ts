import * as Sequelize from "sequelize";
import { Registry } from "./registry";

export type Model = Sequelize.Model<any, any>;

export const assertSequelizeModel = (model: any) => {
  if (model instanceof <any>Sequelize.Model) {
    throw new Error(
      `Expected to receive a Sequelize.Model. Received a sequelize instance instead: ${model}.`
    );
  }
  if (!model || model.constructor !== Sequelize.Model.constructor) {
    throw new Error(
      `Expected to receive a Sequelize.Model. Received ${model} instead.`
    );
  }
};

export const assertRegistry = (registry: any) => {
  if (!(registry instanceof Registry)) {
    throw new Error(
      `Expected to receive a Registry instance. Received ${registry} instead.`
    );
  }
};
