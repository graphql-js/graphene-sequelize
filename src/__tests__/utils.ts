import * as Sequelize from "sequelize";
import { assertSequelizeModel, assertRegistry, Model } from "./../utils";
import { Registry } from "./../registry";

describe("utils", () => {
  const sequelize = new Sequelize("db", "username", "pass", {
    dialect: "sqlite",
    storage: ":memory:",
    logging: () => null
  });

  let model: Model = sequelize.define("user", {
    name: Sequelize.STRING,
    lastName: Sequelize.STRING
  });

  beforeEach(async () => {
    await model.sync();
  });

  afterEach(async () => {
    await model.drop();
  });

  test("assertSequelizeModel fails on non sequelize models", () => {
    expect(() => assertSequelizeModel(null)).toThrowErrorMatchingSnapshot();
  });

  test("assertSequelizeModel fails on model instance", async () => {
    let user = await model.create({ name: "Peter", lastName: "Griffin" });
    expect(() => assertSequelizeModel(user)).toThrowErrorMatchingSnapshot();
  });

  test("assertSequelizeModel succeeds on model", () => {
    assertSequelizeModel(model);
  });

  test("assertRegistry fails on non registry", async () => {
    let registry = new Object();
    expect(() => assertRegistry(registry)).toThrowErrorMatchingSnapshot();
  });

  test("assertRegistry succeeds on registry", () => {
    let registry = new Registry();
    assertRegistry(registry);
  });
});
