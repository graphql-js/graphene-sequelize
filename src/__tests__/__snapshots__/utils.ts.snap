// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils assertRegistry fails on non registry 1`] = `"Expected to receive a Registry instance. Received [object Object] instead."`;

exports[`utils assertSequelizeModel fails on model instance 1`] = `"Expected to receive a Sequelize.Model. Received a sequelize instance instead: [object SequelizeInstance:user]."`;

exports[`utils assertSequelizeModel fails on non sequelize models 1`] = `"Expected to receive a Sequelize.Model. Received null instead."`;
